### ISSUE TITLE
XML documentation refactoring for <see>

### ISSUE THREAD

**Issue by miloush** [dotnet/roslyn#21469]
I would like to suggest a quick action that would turn words in XML documentation to `<see cref>` and `<see langword>` respectively, if applicable.

Example:
```C#
class Program
{
    /// <summary>
    /// Initializes a new instance of the Program class.
    /// </summary>
    Program() {}
}
```

Pressing Ctrl+. on the `Program` word in summary would offer to change Program into `<see cref="Program" />`. Similarly, pressing Ctrl+. on `new` or `class` would offer to change it to `<see langword="new" />`.

(related discussion #11489)

**Comment by sharwell** [comment]
I was just thinking about this!

However, rather than a <kbd>Ctrl</kbd>+<kbd>.</kbd>, I was thinking of tying at least the keyword handling into completion (<kbd>Ctrl</kbd>+<kbd>Space</kbd>).

**Comment by CyrusNajmabadi** [comment]
Work in progress:

![image](https://user-images.githubusercontent.com/4564579/29244599-069ec764-7f71-11e7-9e3a-cf52450e25cb.png)

**Comment by miloush** [comment]
@sharwell I saw your comment on #11489 but I think I prefer the standard refactoring ways to do it, because I wanted 1) to have the ability to fix the whole document 2) see squiggles on potential words that should be <see> things.

**Comment by miloush** [comment]
Hmmmmm looks like I attempted this myself in 2014 :-)

- Apparently I was worried about the performance of checking every word against known symbol and wanted to use it for unknown ones too, so I just offered it on every XML word starting with a capital letter. (I am aware that is against both my arguments to @sharwell above.)
- In addition to that, I also wanted to offer `paramref` and `typeparamref` which I forgot to mention in this issue, but I think that should be in.
- I offered both `see` and `seealso`
- I also wanted to make sure this is the first refactoring in the list, so that you can blindly do Ctrl+. and Enter in the XML comment.
- I obviously didn't know a thing about writing refactoring providers :-)

```C#
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeRefactorings;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
using Microsoft.CodeAnalysis.Text;

namespace XmlDocRefactor
{
    [ExportCodeRefactoringProvider(RefactoringId, LanguageNames.CSharp)]
    [ExtensionOrder(Before = "Change Signature Code Action Provider")]
    internal class CodeRefactoringProvider : ICodeRefactoringProvider
    {
        public const string RefactoringId = "XmlDocRefactor";

        private const string See = "see";
        private const string SeeAlso = "seealso";
        private const string Parameter = "paramref";
        private const string TypeParameter = "typeparamref";
        private const string Keyword = "c";

        private const string CrefAttributeName = "cref";
        private const string NameAttributeName = "name";

        public async Task<IEnumerable<CodeAction>> GetRefactoringsAsync(Document document, TextSpan cursorSpan, CancellationToken cancellationToken)
        {
            SyntaxNode root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);

            XmlTextSyntax textNode = root.FindNode(cursorSpan, true, true) as XmlTextSyntax;
            if (textNode == null)
                return null;

            SyntaxToken textToken = root.FindToken(cursorSpan.Start, true);
            if (textToken.CSharpKind() != SyntaxKind.XmlTextLiteralToken)
                return null;

            TextSpan enclosedSpan = FindIdentifier(document, textToken, cursorSpan);
            if (enclosedSpan.Length < 1)
                return null;

            SourceText documentSource = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
            string enclosed = documentSource.ToString(enclosedSpan);

            if (!SyntaxFacts.IsValidIdentifier(enclosed))
                return null;

            return AvailableActions(document, documentSource, root, cursorSpan, textNode, textToken, enclosedSpan, enclosed);
        }

        private IEnumerable<CodeAction> AvailableActions(Document document, SourceText documentSource, SyntaxNode root, TextSpan cursorSpan, XmlTextSyntax textNode, SyntaxToken textToken, TextSpan enclosedSpan, string enclosed)
        {
            List<CodeAction> actions = new List<CodeAction>(3);
            bool disqualifyVision = false;

            MethodDeclarationSyntax annotatedMethod = root.FindNode(cursorSpan).FirstAncestorOrSelf<MethodDeclarationSyntax>();
            if (annotatedMethod != null)
            {
                if (null != ParameterSyntax parameter = FindParameterIdentifier(annotatedMethod, enclosed))
                    actions.Add(CodeAction.Create(ActionName(enclosed, Parameter), ChangedDocument(document, documentSource, root, textNode, textToken, enclosedSpan, enclosed, Parameter, NameAttributeName)));

                if (null != TypeParameterSyntax typeParameter = FindTypeParameterIdentifier(annotatedMethod, enclosed))
                {
                    actions.Add(CodeAction.Create(ActionName(enclosed, TypeParameter), ChangedDocument(document, documentSource, root, textNode, textToken, enclosedSpan, enclosed, TypeParameter, NameAttributeName)));
                    disqualifyVision = true;
                }
            }

            // symbol lookup takes lot of time and non-existent identifiers might want to be linked anyway
            if (!disqualifyVision && char.IsUpper(enclosed, 0))
            {
                actions.Add(CodeAction.Create(ActionName(enclosed, See), ChangedDocument(document, documentSource, root, textNode, textToken, enclosedSpan, enclosed, See, CrefAttributeName)));
                actions.Add(CodeAction.Create(ActionName(enclosed, SeeAlso), ChangedDocument(document, documentSource, root, textNode, textToken, enclosedSpan, enclosed, SeeAlso, CrefAttributeName)));
            }

            enclosed = enclosed.ToLowerInvariant(); // TODO: language dependent
            if (SyntaxFacts.GetKeywordKind(enclosed) != SyntaxKind.None)
                actions.Add(CodeAction.Create("Link to keyword", ChangedDocument(document, documentSource, root, textNode, textToken, enclosedSpan, enclosed, Keyword)));

            return actions;
        }

        // Before:                                                             After:
        //                                                                    
        //   XmlElementStartTag                                                  XmlElementStartTag
        //   XmlText                                          textNode           XmlText                              textNodes[0]
        //    ┕ XmlTextLiteralToken ('start enclosed end')    textToken          ┕ XmlTextLiteralToken ('start )     textTokenStart
        //   XmlElementEndTag                                                  * Xml(Empty)Element (enclosed)         textNodes[1]
        //                                                                     * XmlText                              textNodes[2]
        //                                                                        ┕ XmlTextLiteralToken ( end')       textTokenEnd  
        //                                                                       XmlElementEndTag

        private Document ChangedDocument(Document document, SourceText documentSource, SyntaxNode root, XmlTextSyntax textNode, SyntaxToken textToken, TextSpan enclosedSpan, string enclosed, string elementName, string attribute = null)
        {
            Split(documentSource, textToken, enclosedSpan, out SyntaxToken textTokenStart, out SyntaxToken textTokenEnd);

            SyntaxNode newTextNode = textNode.ReplaceToken(textToken, new[] { textTokenStart, CreateEnclosedToken(elementName, attribute, enclosed), textTokenEnd });

            //SyntaxNode[] textNodes =
            //{
            //    textNode.ReplaceToken(textToken, textTokenStart).WithTrailingTrivia(),

            //    CreateEnclosedNode(elementName, attribute, enclosed),

            //    XmlText(TokenList(textTokenEnd)).WithTrailingTrivia(textNode.GetTrailingTrivia())
            //};

            SyntaxNode parent = textNode.Parent;

            // SyntaxNode newParent = parent.ReplaceNode(textNode, textNodes);
            SyntaxNode newParent = parent.ReplaceNode(textNode, newTextNode);

            return document.WithSyntaxRoot(root.ReplaceNode(parent, newParent));
        }

        private void Split(SourceText documentSource, SyntaxToken textToken, TextSpan enclosedSpan, out SyntaxToken textTokenStart, out SyntaxToken textTokenEnd)
        {
            Debug.Assert(textToken.IsKind(SyntaxKind.XmlTextLiteralToken));

            string start = documentSource.ToString(TextSpan.FromBounds(textToken.SpanStart, enclosedSpan.Start));
            textTokenStart = XmlTextLiteral(textToken.LeadingTrivia, start, start, TriviaList());

            string end = documentSource.ToString(TextSpan.FromBounds(enclosedSpan.End, textToken.Span.End));
            textTokenEnd = XmlTextLiteral(TriviaList(), end, end, textToken.TrailingTrivia);
        }

        private SyntaxToken CreateEnclosedToken(string elementName, string attribute, string enclosed)
        {
            string xml = string.Format(attribute == null ? "<{0}>{2}</{0}>" : "<{0} {1}=\"{2}\" />", elementName, attribute, enclosed);

            return XmlTextLiteral(TriviaList(), xml, xml, TriviaList());
        }
        private SyntaxNode CreateEnclosedNode(string elementName, string attribute, string enclosed)
        {
            if (attribute == null)

                return XmlElement
                       (
                           XmlElementStartTag(XmlName(elementName)),
                              SingletonList<XmlNodeSyntax>(XmlText(TokenList(XmlTextLiteral(TriviaList(), enclosed, enclosed, TriviaList())))),
                           XmlElementEndTag(XmlName(Keyword))
                       );

            else

                return XmlEmptyElement
                       (
                            XmlName(elementName),
                            SingletonList<XmlAttributeSyntax>
                            (
                                XmlTextAttribute(XmlName(attribute).WithLeadingTrivia(TriviaList(Space)),
                                    Token(SyntaxKind.DoubleQuoteToken),
                                        TokenList(XmlTextLiteral(TriviaList(), enclosed, enclosed, TriviaList())),
                                    Token(SyntaxKind.DoubleQuoteToken))
                            )
                        );
        }

        private TextSpan FindIdentifier(Document document, SyntaxToken xmlToken, TextSpan textSpan)
        {
            if (textSpan.Length > 0)
                return textSpan;

            string tokenText = xmlToken.Text;

            int startIndex = textSpan.Start - xmlToken.Span.Start;

            while (startIndex > 0)
                if (char.IsLetterOrDigit(tokenText, startIndex - 1) || tokenText[startIndex - 1] == '_')
                    --startIndex;
                else
                    break;

            int endIndex = textSpan.End - xmlToken.Span.Start;

            while (endIndex < tokenText.Length - 1)
                if (char.IsLetterOrDigit(tokenText, endIndex + 1) || tokenText[startIndex + 1] == '_')
                    ++endIndex;
                else
                    break;

            return new TextSpan(xmlToken.SpanStart + startIndex, endIndex - startIndex + 1);
        }
        private ParameterSyntax FindParameterIdentifier(MethodDeclarationSyntax method, string identifierText)
        {
            if (method.ParameterList != null)
                foreach (ParameterSyntax parameter in method.ParameterList.Parameters)
                    if (parameter.Identifier.Text == identifierText)
                        return parameter;

            return null;
        }
        private TypeParameterSyntax FindTypeParameterIdentifier(MethodDeclarationSyntax method, string identifierText)
        {
            if (method.TypeParameterList != null)
                foreach (TypeParameterSyntax parameter in method.TypeParameterList.Parameters)
                    if (parameter.Identifier.Text == identifierText)
                        return parameter;

            return null;
        }

        private static string ActionName(string enclosed, string elementName)
        {
            return "Link to '" + enclosed + "' (" + elementName + ")";
        }

        private static IEnumerable<SyntaxNode> ChildNodesBefore(SyntaxNode root, SyntaxNode nodeInList)
        {
            return root.ChildNodes().TakeWhile(node => node != nodeInList);
        }
        private static IEnumerable<SyntaxNode> ChildNodesAfter(SyntaxNode root, SyntaxNode nodeInList)
        {
            return root.ChildNodes().SkipWhile(node => node != nodeInList).Skip(1);
        }
    }
}
```

**Comment by sharwell** [comment]
@miloush This is being implemented for 15.5 as both a refactoring (#21477) and an enhancement to code completion (#21479). It won't have fix-all support, but in my experience the completion behavior covers you quite well - for the words you mentioned it will automatically commit the completion with the single keystroke.
